onmt/models/model.py

```diff
@@ -43,6 +44,63 @@ def update_dropout(self, dropout, attention_dropout):
     def count_parameters(self, log=print):
         raise NotImplementedError
 
+    def load_state_dict(
+        self,
+        checkpoint,
+        precision=torch.float32,
+        device=torch.device("cpu"),
+        strict=True,
+    ):
+        """Custom state_dict loading to enable moving module on device as they are loaded
+
+        Args:
+            checkpoint:
+            precision:
+            device:
+            strict:
+
+        Return:
+
+            * Model"""
+
+        # bitsandbytes quantize weights when .cuda() is called
+        # for huge models we need to save Ram
+        # so we load the weights  module by module and transfer them to GPU for quantization
+
+        for name, module in self.named_modules():
+            for param_name, param in module.named_parameters():
+                if len(param_name.split(".")) == 1:  # only last key
+                    if name + "." + param_name in checkpoint["model"].keys():
+                        param.data = checkpoint["model"][name + "." + param_name]
+                        del checkpoint["model"][name + "." + param_name]
+                    elif (name + "." + param_name).replace(
+                        ".linear.weight", ".weight"
+                    ) in checkpoint["model"].keys():
+                        param.data = checkpoint["model"][
+                            (name + "." + param_name).replace(
+                                ".linear.weight", ".weight"
+                            )
+                        ]
+                        del checkpoint["model"][
+                            (name + "." + param_name).replace(
+                                ".linear.weight", ".weight"
+                            )
+                        ]
```

Comment:
if we retain QLoraLinear and remove QQLoraLinear then thsi will be way less cryptic because we won't have to remap the stuff like linear_key(value,query).linear.weight to linear_key(value,query).weight